exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/OutputWidthVisitor.java

```diff
@@ -0,0 +1,278 @@
+     * a {@link FixedLenExpr} by taking the max of the if-expr-width and the else-expr-width.
+     *
+     * @param ifExpression
+     * @param state
+     * @return IfElseWidthExpr
+     * @throws RuntimeException
+     */
+    @Override
+    public OutputWidthExpression visitIfExpression(IfExpression ifExpression, OutputWidthVisitorState state)
+                                                                    throws RuntimeException {
+        IfExpression.IfCondition condition = ifExpression.ifCondition;
+        LogicalExpression ifExpr = condition.expression;
+        LogicalExpression elseExpr = ifExpression.elseExpression;
+
+        OutputWidthExpression ifWidthExpr = ifExpr.accept(this, state);
+        OutputWidthExpression elseWidthExpr = null;
+        if (elseExpr != null) {
+            elseWidthExpr = elseExpr.accept(this, state);
+        }
+        return new IfElseWidthExpr(ifWidthExpr, elseWidthExpr);
+    }
+
+    /**
+     * Handles a {@link FunctionHolderExpression}. Functions that produce fixed-width output are trivially
+     * converted to a {@link FixedLenExpr}. For functions that produce variable width output, the output width calculator
+     * annotation is looked-up and recorded in a {@link FunctionCallExpr}. This calculator will later be used to convert
+     * the FunctionCallExpr to a {@link FixedLenExpr} expression
+     * @param holderExpr
+     * @param state
+     * @return FunctionCallExpr
+     * @throws RuntimeException
+     */
+    @Override
+    public OutputWidthExpression visitFunctionHolderExpression(FunctionHolderExpression holderExpr,
+                                                               OutputWidthVisitorState state) throws RuntimeException {
+        OutputWidthExpression fixedWidth = getFixedLenExpr(holderExpr.getMajorType());
+        if (fixedWidth != null) { return fixedWidth; }
+        // Only Drill functions can be handled. Non-drill Functions, like HiveFunctions
+        // will default to a fixed value
+        if (!(holderExpr instanceof DrillFuncHolderExpr)) {
+            // We currently only know how to handle DrillFuncs.
+            // Use a default if this is not a DrillFunc
+            return new FixedLenExpr(OutputSizeEstimateConstants.NON_DRILL_FUNCTION_OUTPUT_SIZE_ESTIMATE);
+        }
+
+        final DrillFuncHolder holder = ((DrillFuncHolderExpr) holderExpr).getHolder();
+
+        // If the user has provided a size estimate, use it
+        int estimate = holder.variableOuputSizeEstimate();
+        if (estimate != FunctionTemplate.VARIABLE_OUTPUT_SIZE_ESTIMATE_DEFAULT) {
+            return new FixedLenExpr(estimate);
+        }
+        // Use the calculator provided by the user or use the default
+        OutputWidthCalculator widthCalculator = holder.getOutputWidthCalculator();
+        final int argSize = holderExpr.args.size();
+        ArrayList<OutputWidthExpression> arguments = null;
+        if (argSize != 0) {
```

Comment:
Trivial question: Would it also work without that if() check ?  I.e., use an empty ArrayList in case of zero args.